test/concurrent/WaitLockTest.ooc

```diff
@@ -0,0 +1,113 @@
+/* This file is part of magic-sdk, an sdk for the open source programming language magic.
+ *
+ * Copyright (C) 2016 magic-lang
+ *
+ * This software may be modified and distributed under the terms
+ * of the MIT license.  See the LICENSE file for details.
+ */
+
+use concurrent
+use unit
+
+WaitLockTest: class extends Fixture {
+	init: func {
+		super("WithLock")
+		this add("_testWithMutexOwnership", This _testWithMutexOwnership)
+		this add("_testWithoutMutexOwnership", This _testWithoutMutexOwnership)
+		this add("_testWakeWithFailingCondition", This _testWakeWithFailingCondition)
+		this add("_testWakeWithPassingCondition", This _testWakeWithPassingCondition)
+	}
+	_testWithMutexOwnership: static func {
+		waitLock := WaitLock new()
+		waitLock free()
+	}
+	_testWithoutMutexOwnership: static func {
+		mutex := Mutex new()
+		waitLock := WaitLock new(mutex)
+		waitLock free()
+		mutex free()
+	}
+	_testWakeWithFailingCondition: static func {
+		timesTriggered := 0
+		timesTriggeredRef := timesTriggered&
+		waitLock := WaitLock new()
+		waitingThread := Thread new(||
+			waitLock lockWhen(||
+				timesTriggeredRef@ = timesTriggeredRef@ + 1
```

Comment:
Probably some confusion regarding closure captures?  
rock should be able to mark and capture `timesTriggered` by reference if the line is changed to `timesTriggered = timesTriggered + 1`  

@erikhagglund If rock was not able to do that, it should be reported as a bug, otherwise just using the original variable is probably clearer.